### PR TITLE
Suggest `_` and `..` if a pattern has too few fields

### LINKED ISSUE #80010: diagnostics: mismatched pattern field: suggest ".." instead of "_"
````
error[E0023]: this pattern has 1 field, but the corresponding tuple variant has 3 fields
   --> clippy_lints/src/needless_borrow.rs:50:20
    |
50  |             if let ty::Ref(_) = cx.typeck_results().expr_ty(inner).kind() {
    |                    ^^^^^^^^^^ expected 3 fields, found 1

````
trying to ignore the inner fields with `ty::Ref(_)`  does not work because it has 3 fields.
The message could suggest `Ref(..)` if there are multiple fields and it thinks we don't want to bind any:

````
error[E0023]: this pattern has 1 field, but the corresponding tuple variant has 3 fields
   --> clippy_lints/src/needless_borrow.rs:50:20
    |
50  |             if let ty::Ref(_) = cx.typeck_results().expr_ty(inner).kind() {
    |                    ^^^^^^^^^^ expected 3 fields, found 1
    = help: Use `ty::Ref(..)` to ignore any inner fields
````

### PR DESCRIPTION
**Pull request by camelid**
Fixes #80010.

### CODE REVIEW

File: src/test/ui/pattern/pat-tuple-underfield.fixed [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// ignore-tidy-filelength
+
+struct S(i32, f32);
+enum E {
+    S(i32, f32),
+}
+
+fn main() {
+    match S(0, 1.0) {
+        S(x).. => {}
```
Comment (camelid) [inline src/test/ui/pattern/pat-tuple-underfield.fixed:11]: See https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Span.3A.3Ashrink_to_hi.20-.201.

File: src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr [hunk 2]
```diff
@@ -31,6 +31,8 @@ LL | struct TupleStruct<S, T>(S, T);
 ...
 LL |     TupleStruct(_) = TupleStruct(1, 2);
    |     ^^^^^^^^^^^^^^ expected 2 fields, found 1
+   |
+   = help: use `..` to ignore unmentioned fields
```
Comment (pickfire) [inline src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr:35]: I think it should also mention another way which is to use `_, _` to fill out all the fields, this is more explicit and I think should be more preferable.

File: src/test/ui/pattern/pat-tuple-underfield.fixed [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        //~| HELP use `..` to ignore all unmentioned fields
+    }
+    match E::S(0, 1.0) {
+        E::S(_, .., _) => {}
+        //~^ ERROR this pattern has 1 field, but the corresponding tuple variant has 2 fields
+        //~| HELP use `_` to explicitly ignore a field
+        //~| HELP use `..` to ignore all unmentioned fields
+    }
+    match E::S(0, 1.0) {
+        E::S(.._, _) => {}
```
Comment (camelid) [inline src/test/ui/pattern/pat-tuple-underfield.fixed:41]: As you can see, unfortunately it is applying both suggestions. Is there any way to signal that only one should be applied by rustfix?

Comment (camelid) [reply src/test/ui/pattern/pat-tuple-underfield.fixed]: I removed the test since it doesn't seem to be possible. I'm leaving this conversation unresolved in case someone knows a way to do it!

File: src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr [hunk 4]
```diff
@@ -49,6 +58,15 @@ LL |     SingleVariant(S, T)
...
 LL |     Enum::SingleVariant(_) = Enum::SingleVariant(1, 2);
    |     ^^^^^^^^^^^^^^^^^^^^^^ expected 2 fields, found 1
+   |
+help: use `_` to explicitly ignore each field
+   |
+LL |     Enum::SingleVariant(_, _) = Enum::SingleVariant(1, 2);
+   |                          ^^^
+help: use `..` to ignore all unmentioned fields
+   |
+LL |     Enum::SingleVariant(_, ..) = Enum::SingleVariant(1, 2);
```
Comment (pickfire) [inline src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr:68]: I believe the other method is to use `(..)` but not `(_, ..)` which is a mixture of both.

File: compiler/rustc_typeck/src/check/pat.rs [hunk 5]
```diff
@@ -1040,6 +1042,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 vec![(left, "(".to_string()), (right.shrink_to_hi(), ")".to_string())],
                 Applicability::MachineApplicable,
             );
+        } else if fields.len() > subpats.len() {
+            let after_fields_span = if pat_span == DUMMY_SP {
+                pat_span
```
Comment (pickfire) [inline compiler/rustc_typeck/src/check/pat.rs:1047]: What is the `DUMMY_SP` for and this branch for? I never seen this before.

File: compiler/rustc_typeck/src/check/pat.rs [hunk 6]
```diff
@@ -1040,6 +1042,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                "use `_` to explicitly ignore each field",
+                wildcard_sugg,
+                Applicability::MaybeIncorrect,
+            );
+            err.span_suggestion(
+                after_fields_span,
+                "use `..` to ignore all unmentioned fields",
+                rest_sugg,
+                Applicability::MaybeIncorrect,
+            );
```
Comment (pickfire) [inline compiler/rustc_typeck/src/check/pat.rs:1071]: I believe you might want to use `span_suggestions` to suggest multiple items, rustfix checks this but maybe not in our case. https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic/struct.Diagnostic.html#method.span_suggestions https://github.com/rust-lang/rustfix/issues/78

```suggestion
            err.span_suggestions(
                after_fields_span,
                "use `_` to explicitly ignore each field or `..` to ignore all unmentioned fields",
                vec![wildcard_sugg, rest_sugg].into_iter(),
                Applicability::MaybeIncorrect,
            );
```
I wonder if `vec![wildcard_sugg, rest_sugg].into_iter(),` could be improved?

But we can even leave a `FIXME` mentioning where it could improve the applicability to a more correct one with some type matching to know which position to fill with `_`.

File: src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr [hunk 7]
```diff
@@ -17,6 +17,15 @@ LL | struct P<T>(T); // 1 type parameter wanted
...
    |         ^^^ expected 1 field, found 0
+   |
+help: use `_` to explicitly ignore each field
+   |
+LL |     let P(_) = U {};
+   |           ^
+help: use `..` to ignore all unmentioned fields
+   |
+LL |     let P(..) = U {};
+   |           ^^
```
Comment (pickfire) [inline src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr:28]: Or maybe we shouldn't even show this up when there is one field so rustfix can apply the suggestion?

File: compiler/rustc_typeck/src/check/pat.rs (hunk 5 above)
Comment (camelid) [reply compiler/rustc_typeck/src/check/pat.rs:1047]: This is to prevent a compiler crash if `pat_span == DUMMY_SP` since we then subtract `1` from `pat_span` and `DUMMY_SP` is basically equal to 0.

File: compiler/rustc_typeck/src/check/pat.rs (hunk 6 above)
Comment (camelid) [reply compiler/rustc_typeck/src/check/pat.rs:1071]: I like this idea, but unfortunately I don't think the loss in readability is worth it. Currently, the output looks like this:

```
help: use `_` to explicitly ignore each field
   |
LL |         E::S(_, _) => {}
   |              ^^^^
help: use `..` to ignore all unmentioned fields
   |
LL |         E::S(..) => {}
   |              ^^
```

But using your suggestion makes the output look like this:

```
help: use `_` to explicitly ignore each field or `..` to ignore all unmentioned fields
   |
39 |         E::S(_, _) => {}
   |              ^^^^
39 |         E::S(..) => {}
   |              ^^
```

Which is a bit overwhelming and not as clear for the user. Perhaps we can make changes to the compiler's error system to make multiple suggestions easier to read for the user.

There is precedent for using multiple calls to `span_suggestion` to make the output easier to read:

```
error: unexpected `,` in pattern
 --> src/main.rs:2:10
  |
2 |     let x, y = 3;
  |          ^
  |
help: try adding parentheses to match on a tuple...
  |
2 |     let (x, y) = 3;
  |         ^^^^^^
help: ...or a vertical bar to match on multiple alternatives
  |
2 |     let x | y = 3;
  |         ^^^^^
```

File: src/test/ui/pattern/pat-tuple-underfield.fixed (hunk 3 above)
Comment (estebank) [reply src/test/ui/pattern/pat-tuple-underfield.fixed]: This is a dual issue: we need to change rustc's json output to be more detailed (it cannot distinguish today between multiple one part suggestions and a single multipart suggestion) and rustfix needs to start consuming it. We started talking about this a year and a half ago, but it petered out, in part due to lack of time on my side. I'm hoping to rekindle this and start fixing this in the first half of 2021.

File: compiler/rustc_typeck/src/check/pat.rs (hunk 5 above)
Comment (estebank) [reply compiler/rustc_typeck/src/check/pat.rs:1047]: If `pat_span` is `DUMMY_SP` then you shouldn't be giving suggestions at all.

File: compiler/rustc_typeck/src/check/pat.rs [hunk 8]
```diff
@@ -1040,6 +1042,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
             );
+        } else if fields.len() > subpats.len() {
+            let after_fields_span = if pat_span == DUMMY_SP {
+                pat_span
+            } else {
+                pat_span.with_hi(pat_span.hi() - BytePos(1)).shrink_to_hi()
+            };
+
+            let mut wildcard_sugg =
+                iter::repeat("_").take(fields.len() - subpats.len()).collect::<Vec<_>>().join(", ");
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/pat.rs:1053]:
```suggestion
            let mut wildcard_sugg =vec!["_"; fields.len() - subpats.len()].join(", ");
```

File: src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr [hunk 9]
```diff
@@ -49,6 +58,15 @@ LL |     SingleVariant(S, T)
...
+   |                          ^^^^
```
Comment (estebank) [inline src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr:69]: I think that we should check that the present fields are all already placeholders, and if so go ahead and suggest `Enum::SingleVariant(..)` instead here.

File: src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr (hunk 7 above)
Comment (estebank) [reply src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr:28]: I agree, we shouldn't suggest `..` when there's only a single field left (including when the current code is `P(a)` and the suggestion is for `P(a, _)`, but not if the current code is `P(_)`, then the suggestion should be for `P(..)`).

File: src/test/ui/pattern/pat-tuple-underfield.fixed (hunk 3 above)
Comment (camelid) [reply src/test/ui/pattern/pat-tuple-underfield.fixed]: That would be amazing! ❤️

**Comment by estebank** [comment]
Sorry for the delay in reviewing. I left a few nitpicks that require changing the logic slightly. Would you mind incorporating those changes and pinging me once that's done?

**Comment by camelid** [comment]
> Sorry for the delay in reviewing.

That's okay; it seems most of the diagnostics PRs get assigned to you :)

> I left a few nitpicks that require changing the logic slightly. Would you mind incorporating those changes and pinging me once that's done?

👍

File: compiler/rustc_typeck/src/check/pat.rs (hunk 5 above)
Comment (camelid) [reply compiler/rustc_typeck/src/check/pat.rs:1047]: What happens if the error rendering system receives a diagnostic that has `DUMMY_SP`? Will it just ignore it?

File: src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr (hunk 9 above)
Comment (camelid) [reply src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr]: Done.

File: compiler/rustc_typeck/src/check/pat.rs (hunk 6 above)
Comment (pickfire) [reply compiler/rustc_typeck/src/check/pat.rs]: > Which is a bit overwhelming and not as clear for the user. Perhaps we can make changes to the compiler's error system to make multiple suggestions easier to read for the user.

Yes, this is a better option but we should make rustfix aware of this, if you add it please add it to rustfix too.

File: compiler/rustc_typeck/src/check/pat.rs [hunk 10]
```diff
@@ -1029,17 +1030,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 // last sub-pattern. In the case of `A(x)` the first and last may coincide.
                 // This looks like:
                 //
-                // help: missing parenthesis
+                // help: missing parentheses
```
Comment (pickfire) [inline compiler/rustc_typeck/src/check/pat.rs:1033]: This should be `parenthesis`, not `parentheses`.

File: compiler/rustc_typeck/src/check/pat.rs [hunk 11]
```diff
@@ -1029,17 +1030,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
                 //   |
                 // L |     let A((x, y)) = A((1, 2));
                 //   |           ^    ^
                 [first, ..] => (first.span.shrink_to_lo(), subpats.last().unwrap().span),
             };
             err.multipart_suggestion(
-                "missing parenthesis",
+                "missing parentheses",
```
Comment (pickfire) [inline compiler/rustc_typeck/src/check/pat.rs:1040]:
```suggestion
                "missing parenthesis",
```
Unnecessary change, it became a typo.

File: compiler/rustc_typeck/src/check/pat.rs [hunk 12]
```diff
@@ -1029,17 +1030,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                    Applicability::MaybeIncorrect,
+                );
+            } else {
+                err.span_suggestion(
+                    after_fields_span,
+                    "use `..` to ignore the rest of the fields",
+                    String::from(", .."),
+                    Applicability::MaybeIncorrect,
+                );
+            }
```
Comment (pickfire) [inline compiler/rustc_typeck/src/check/pat.rs:1085]: We should create a helper function for `span_suggestions` but with different message, and at the same time open an issue in rustfix that this exists so they can integrate with it.

File: compiler/rustc_typeck/src/check/pat.rs (hunk 11 above)
Comment (camelid) [reply compiler/rustc_typeck/src/check/pat.rs:1040]: No, it was a typo _before_.

File: compiler/rustc_typeck/src/check/pat.rs (hunk 12 above)
Comment (camelid) [reply compiler/rustc_typeck/src/check/pat.rs:1085]: I’m not sure what you’re suggesting. Can you please explain what you mean?

Comment (pickfire) [reply compiler/rustc_typeck/src/check/pat.rs:1085]: I similar helper to `span_suggestion` so we don't need to call it multiple times, as well as being recognized by rustfix, not sure if rustfix is aware that multiple suggestions can be used on the same line right now.

Comment (camelid) [reply compiler/rustc_typeck/src/check/pat.rs:1085]: See https://github.com/rust-lang/rust/pull/80017#discussion_r546165651.

This doesn’t make sense to do in this PR.

File: src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr (hunk 7 above)
Comment (camelid) [reply src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr]: I would rather not make this change too complex. Do you think we could do that in a followup PR?

**Comment by camelid** [comment]
> Sorry for the delay in reviewing. I left a few nitpicks that require changing the logic slightly. Would you mind incorporating those changes and pinging me once that's done?

@estebank here's your ping!

I didn't make one of the changes because of https://github.com/rust-lang/rust/pull/80017#discussion_r546446547.

File: src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr (hunk 7 above)
Comment (pickfire) [reply src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr]: No, I believe it is simple enough to do (a simple one) in this PR. Let me send a suggestion for that.

File: compiler/rustc_typeck/src/check/pat.rs (hunk 12 above)
Comment (pickfire) [inline compiler/rustc_typeck/src/check/pat.rs:1085]: Suggesting `..` only when there are more than one `_` is simple enough to be in this PR.
```suggestion
            // Only suggest `..` if there are more than one `_`. 
            if fields.len() - subpats.len() > 1 {            
                if subpats.is_empty() || all_wildcards {     
                    err.span_suggestion(                     
                        all_fields_span,                     
                        "use `..` to ignore all fields",     
                        String::from(".."),             
                        Applicability::MaybeIncorrect,  
                    );                                  
                } else {                           
                    err.span_suggestion(           
                        after_fields_span,
                        "use `..` to ignore the rest of the fields",
                        String::from(", .."),                                 
                        Applicability::MaybeIncorrect,                        
                    );                                                        
                }
            }                                                                 
```

File: compiler/rustc_typeck/src/check/pat.rs [hunk 13]
```diff
@@ -1029,17 +1030,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            let all_fields_span = match subpats {
+                [] => after_fields_span,
+                [field] => field.span,
+                [first, .., last] => first.span.to(last.span),
+            };
+
+            // Check if all the fields in the pattern are wildcards.
+            let all_wildcards = subpats.iter().all(|pat| matches!(pat.kind, PatKind::Wild));
+
+            let mut wildcard_sugg = vec!["_"; fields.len() - subpats.len()].join(", ");
```
Comment (pickfire) [inline compiler/rustc_typeck/src/check/pat.rs:1059]: Would be good to add a comment for potential improvements here.

```suggestion
            // FIXME: heuristic-based suggestion to check current types for where to add `_`.
            let mut wildcard_sugg = vec!["_"; fields.len() - subpats.len()].join(", ");
```

File: compiler/rustc_typeck/src/check/pat.rs (hunk 5 above)
Comment (estebank) [reply compiler/rustc_typeck/src/check/pat.rs:1046]: The error renderer will ignore `DUMMY_SP`s and not show labels associated to it, while still showing messages, but they do have the potential to break suggestions and _certainly_ break everything when synthesizing a new span from them.

File: compiler/rustc_typeck/src/check/pat.rs (hunk 12 above)
Comment (estebank) [reply compiler/rustc_typeck/src/check/pat.rs:1085]: There's a precondition to doing this which is reworking the json output schema which I will finally get around to doing this year. It won't be addressable in this PR as it is a ton of unrelated work.

File: src/test/ui/issues/issue-72574-2.stderr [hunk 14]
```diff
@@ -26,6 +26,15 @@ LL | struct Binder(i32, i32, i32);
...
    |         ^^^^^^^^^^^^^^^^^^^ expected 3 fields, found 2
+   |
+help: use `_` to explicitly ignore each field
+   |
+LL |         Binder(_a, _x @ .., _) => {}
+   |                           ^^^
+help: use `..` to ignore the rest of the fields
+   |
+LL |         Binder(_a, _x @ .., ..) => {}
+   |                           ^^^^
```
Comment (estebank) [inline src/test/ui/issues/issue-72574-2.stderr:37]: This is a pretty funky test case, isn't it? 😅

File: compiler/rustc_typeck/src/check/pat.rs [hunk 15]
```diff
@@ -1029,17 +1030,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                "missing parenthesis",
+                "missing parentheses",
                 vec![(left, "(".to_string()), (right.shrink_to_hi(), ")".to_string())],
                 Applicability::MachineApplicable,
             );
+        } else if fields.len() > subpats.len() {
+            let after_fields_span = if pat_span == DUMMY_SP {
+                pat_span
+            } else {
+                pat_span.with_hi(pat_span.hi() - BytePos(1)).shrink_to_hi()
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/pat.rs:1048]: Can you test whether this breaks in the case of a pattern like `Foo(a    ,    _    )` where `Foo` has 3 fields?

**Comment by estebank** [comment]
I think this PR is in a good state to be merged, I have only one small concern that we might want to punt on, and potentially address @pickfire's comments, but otherwise r=me.

File: compiler/rustc_typeck/src/check/pat.rs (hunk 15 above)
Comment (camelid) [reply compiler/rustc_typeck/src/check/pat.rs:1048]: Are the spaces inside the parentheses significant? Or do you mean `Foo(a, _)`?

Comment (estebank) [reply compiler/rustc_typeck/src/check/pat.rs:1048]: They are significant. I mean to test a case a case where there are multiple spaces between `(`, `a`, `,`, `_` and `)`:

```rust
Foo(   a   ,     _    )
```

File: src/test/ui/error-codes/E0023.stderr [hunk 16]
```diff
@@ -5,16 +5,10 @@ LL |     Apple(String, String),
...
-LL |         Fruit::Apple(a, _) => {},
-   |                       ^^^
-help: use `..` to ignore the rest of the fields
-   |
-LL |         Fruit::Apple(a, ..) => {},
-   |                       ^^^^
+   |         ^^^^^^^^^^^^^^-
+   |         |             |
+   |         |             help: use `_` to explicitly ignore each field
+   |         expected 2 fields, found 1
```
Comment (camelid) [inline src/test/ui/error-codes/E0023.stderr]: Ideally this wouldn't show the `help` inline. Is there any way to force it to be how it was before?

Comment (estebank) [reply src/test/ui/error-codes/E0023.stderr]: `span_suggestion_verbose` always displays the suggestion in its own subdiagnostic window.

**Comment by pickfire** [comment]
I have seen test cases that involves rustfix in the rust repository, maybe we can try adding a sample test case to see if one of the multiple suggestions gets applied simultaneously or only one suggestion?

File: src/test/ui/error-codes/E0023.stderr (hunk 16 above)
Comment (camelid) [reply src/test/ui/error-codes/E0023.stderr]: Ah, I didn't realize that! The docs just say:

> Always show the suggested change.

I didn't realize that meant "always show in its own window". Do you think you could update the docs? If not, I can perhaps do it.

**Comment by camelid** [comment]
I already tried that, and it doesn't work: https://github.com/rust-lang/rust/pull/80017#discussion_r542972452. The messages are for the most part shown in their own diagnostic window which shows the applied suggestion; that should be good enough.

File: compiler/rustc_typeck/src/check/pat.rs (hunk 15 above)
Comment (camelid) [reply compiler/rustc_typeck/src/check/pat.rs:1048]: Added the test. It seems to work.

File: src/test/ui/error-codes/E0023.stderr (hunk 16 above)
Comment (estebank) [reply src/test/ui/error-codes/E0023.stderr]: I can look into updating the docs, there's a lot that could be improved there :)

**Comment by camelid** [comment]
@estebank I think this is ready for a final (or close to final) review!

File: src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr [hunk 17]
```diff
@@ -31,6 +31,11 @@ LL | struct TupleStruct<S, T>(S, T);
 ...
 LL |     TupleStruct(_) = TupleStruct(1, 2);
    |     ^^^^^^^^^^^^^^ expected 2 fields, found 1
+   |
+help: use `_` to explicitly ignore each field
+   |
+LL |     TupleStruct(_, _) = TupleStruct(1, 2);
+   |                  ^^^
```
Comment (estebank) [inline src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr:38]: These two cases should also be suggesting `..` as an alternative, no?

File: src/test/ui/pattern/pat-tuple-underfield.stderr [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 4 fields, found 2
+   |
+help: use `_` to explicitly ignore each field
+   |
+LL |         Point4(   a   ,     _    , _, _) => {}
+   |                                  ^^^^^^
+help: use `..` to ignore the rest of the fields
+   |
+LL |         Point4(   a   ,     _    , ..) => {}
+   |                                  ^^^^
```
Comment (estebank) [inline src/test/ui/pattern/pat-tuple-underfield.stderr:118]: Thanks, this is a great check.

File: src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr (hunk 17 above)
Comment (camelid) [reply src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr:38]: Well, if someone is already explicitly listing out wildcards, should we suggest that they use `..`? On the other hand, someone may not know that `..` is an option and would rather that over listing out wildcards manually. So I think overall I agree with you that we should be giving it as an alternative :)

I'll change it, but for other things we can always adjust the behavior later; I'm hoping to have this PR finished soon.

File: src/test/ui/pattern/pat-tuple-underfield.stderr (hunk 18 above)
Comment (estebank) [inline src/test/ui/pattern/pat-tuple-underfield.stderr:118]: Shouldn't this one be the following?

```

help: use `..` to ignore the rest of the fields
   |
LL |         Point4(   a   , ..) => {}
   |                       ^^^^
```

File: src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr (hunk 17 above)
Comment (estebank) [reply src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr:38]: Sounds reasonable, I don't foresee any big changes. I think that the code itself looks good but want to nail down the exact behavior we want.

Comment (camelid) [reply src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr:38]: So is the behavior you're suggesting that we also suggest `S(..)` if it's currently of the form `S(_, _, _)`?

Comment (estebank) [reply src/test/ui/destructuring-assignment/tuple_struct_destructure_fail.stderr:38]: Yes, I think that if it is currently `S(_, _)` and we suggest `S(_, _, _)` we should also suggest `S(..)`. Having said that, if you find that it will be too much work, only create a ticket and r=me.

File: src/test/ui/pattern/pat-tuple-underfield.stderr (hunk 18 above)
Comment (camelid) [reply src/test/ui/pattern/pat-tuple-underfield.stderr:118]: Ideally yes, but I think that would make the code much more complicated. This is the current behavior:

- If it's of the form `(_, _)` then suggest `(..)`
- Otherwise, suggest `(/* ... */, ..)`

Of course, there's different behavior for the case when only one field is missing :)

File: src/test/ui/pattern/pat-tuple-underfield.fixed (hunk 3 above)
Comment (pickfire) [reply src/test/ui/pattern/pat-tuple-underfield.fixed]: Ah, I did saw this. Did it break or it shows the incorrect behavior? If it's just incorrect behavior we could add it to the tests as well, and once that behavior is fixed we can go and fix the tests.

File: src/test/ui/match/match-pattern-field-mismatch.stderr [hunk 19]
```diff
@@ -6,6 +6,15 @@ LL |         Rgb(usize, usize, usize),
...
 LL |           Color::Rgb(_, _) => { }
    |           ^^^^^^^^^^^^^^^^ expected 3 fields, found 2
+   |
+help: use `_` to explicitly ignore each field
+   |
+LL |           Color::Rgb(_, _, _) => { }
+   |                          ^^^
+help: use `..` to ignore all fields
+   |
+LL |           Color::Rgb(..) => { }
```
Comment (pickfire) [inline src/test/ui/match/match-pattern-field-mismatch.stderr:16]: Why don't we suggest `Color::Rgb(_, _, ..)` as well?

File: compiler/rustc_typeck/src/check/pat.rs [hunk 20]
```diff
@@ -1029,17 +1030,63 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+
+            err.span_suggestion_verbose(
+                after_fields_span,
+                "use `_` to explicitly ignore each field",
+                wildcard_sugg,
+                Applicability::MaybeIncorrect,
+            );
+
+            // Only suggest `..` if more than one field is missing
+            // or the pattern consists of all wildcards.
```
Comment (pickfire) [inline compiler/rustc_typeck/src/check/pat.rs:1072]:
```suggestion
            // or the pattern consists of all wildcards `(_, _)`.
```

Would be better to show an example here for the comment.

File: src/test/ui/match/match-pattern-field-mismatch.stderr (hunk 19 above)
Comment (estebank) [reply src/test/ui/match/match-pattern-field-mismatch.stderr:16]: _I_ consider `Color::Rgb(_, _, ..)` is bad style. :)

File: src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr [hunk 21]
```diff
@@ -17,6 +17,15 @@ LL | struct P<T>(T); // 1 type parameter wanted
...
    |         ^^^ expected 1 field, found 0
+   |
+help: use `_` to explicitly ignore each field
+   |
+LL |     let P(_) = U {};
+   |           ^
+help: use `..` to ignore all fields
+   |
+LL |     let P(..) = U {};
+   |           ^^
```
Comment (estebank) [inline src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr:28]: I think that we'll need to special case when there's only one field missing (and all the other preceding fields aren't `_`) and _only suggest `_`. To clarify, I think that in _this_ case we should _only_ provide the first suggestion, not the `..` one. That being said, it is something I can look at myself if you're itching to merge asap.

It also feels like we could deal with language for a single field. My ideal output for _this_ case would be:

```
help: use `_` to explicitly ignore the missing field
   |
LL |     let P(_) = U {};
   |           ^
```

Comment (camelid) [reply src/test/ui/issues/issue-67037-pat-tup-scrut-ty-diff-less-fields.stderr:28]: > To clarify, I think that in this case we should only provide the first suggestion, not the .. one.

I actually think we _should_ suggest `..` in this case. It's very possible that someone didn't list any fields because they want to match on the outer structure, and so they just want to ignore all of them. We should definitely suggest `..` in that case.

And also, as you said, I would like to get this merged soon. I appreciate the desire to make diagnostics as good as possible - I would like to as well - but I feel that further changes would have diminishing returns and increase the risk of introducing bugs. Also, this change has taken way longer than I anticipated and I would like to spend my time on some other things :)

**Comment by estebank** [comment]
> this change has taken way longer than I anticipated and I would like to spend my time on some other things :)

👍 

@bors r+
